Mitigating performance regressions in 2.0

philippjfr opened the pull request:
An attempt at mitigating some of the performance regressions associated with Param 2.0

# Comparing 2.0.0a3 vs this branch

<img width="953" alt="Screen Shot 2023-09-22 at 15 35 39" src="https://github.com/holoviz/param/assets/1550771/66c50629-3000-4f01-ac87-107b96716411">


# Comparing 1.13.0 vs this branch

<img width="953" alt="Screen Shot 2023-09-22 at 15 36 19" src="https://github.com/holoviz/param/assets/1550771/88ad33c5-02b6-48ae-b16d-b883ffe53695">

philippjfr commented on param/parameterized.py:
```diff
@@ -403,19 +403,14 @@ def iscoroutinefunction(function):
-    p.watchers = {}
-
-    # shallow-copy any mutable slot values other than the actual default
-    for s in p.__class__.__slots__:
```
It seemed a little silly to set slot values twice (once in `__setstate__` and once here), but I think I should revert this.

Hoxbro commented on param/parameterized.py:
```diff
@@ -897,25 +898,35 @@ class ParameterMetaclass(type):
+        all_slots = set()
+        for base in bases:
+            for bcls in base.__mro__:
+                for slot in getattr(bcls, '__slots__', []):
```
I think it would be faster if this `for` statement is skipped by converting directly to a set and updating `all_slots`.

philippjfr commented:
Added a few more micro-optimizations and dropped backward compatibility for unpickling objects from before watchers were introduced (i.e. v1.8.0), which is 5 years ago now.

jbednar commented on param/parameterized.py:
```diff
@@ -403,14 +404,15 @@ def iscoroutinefunction(function):
+        paramobj.watchers = {}
+        p = copy.copy(paramobj)
+    finally:
+        paramobj.watchers = watchers
```
This obscure code coming back deserves a clear comment explaining why it's done this way...

Hoxbro commented on param/parameterized.py:
```diff
@@ -1318,29 +1317,29 @@ def _set_instantiate(self,instantiate):
+                raise AttributeError("Parameter name cannot be modified after "
+                                     "it has been bound to a Parameterized.")
+
+        is_slot = attribute in self.__class__._all_slots_
+        has_watcher = attribute != "default" and attribute in getattr(self, 'watchers', [])
+        if not (is_slot or has_watcher):
```
We could use the walrus operator here so we exit earlier and do not need the calculation of `has_watcher`. 

This is only possible if min. version is Python 3.8.

jbednar replied:
py38 being four years old at this point, I'm happy with that as a version floor for Param 2.0, if that's ok for Panel purposes.

maximlt replied:
@Hoxbro how would you use the walrus operator here?

I believe this could be rewritten as follows. But  I'm happy with how it is right now and haven't run any benchmark to see whether it'd make any difference.

```python
if not is_slot:
    has_watcher = attribute != "default" and attribute in getattr(self, 'watchers', [])
    if not has_watcher:
        # Return early if attribute is not a slot
        return super().__setattr__(attribute, value)
```

Oh and the minimum Python version for Param is 3.8 already :)

Hoxbro replied:
I didn't put that many thoughts into it. 

Looking at this again. I don't think it is possible to do as both variables are needed if we don't exit early.

Your example will give an `UnboundLocalError` and similar with the Walrus operator code change.

maximlt replied:
Ah indeed `has_watcher` is re-used below, my bad!

maximlt reviewed (commented):
LGTM! The numbers look great and some of the refactoring makes the code even more understandable 👍 I just have a question in a comment.

philippjfr commented on param/parameterized.py:
```diff
@@ -897,25 +898,33 @@ class ParameterMetaclass(type):
+            all_slots.update(dict.fromkeys(classdict['__slots__']))
 
-        # No special handling for a __dict__ slot; should there be?
+        classdict['_all_slots_'] = list(all_slots)
```
Fastest way to get an ordered set in python.

maximlt replied:
Thanks that's what I thought but I wanted to be sure, I've added a comment in the code.
